### DISCUSSION TITLE
Thumbnail generation slows down display of pages containing products

### DISCUSSION THREAD

**Discussion by lmeyer1** [PrestaShop/PrestaShop#40175] [Ideas]
#### Is your feature request related to a problem?

I'm always frustrated, when I have to wait a minute or two for a page in Prestashop to display. This happens, when resized product images are missing.

I have tens of thousands of product images. Image resizing via *Image Settings*, *Regenerate Thumbnails* is just impossible. An automated task takes days. On my test deployment, when there are thumbnails missing (maybe I changed image generation settings or thumbnail sizes), a product page may take beween 30 seconds and a couple of minutes to display.

Reason: In class `ImageRetriever`, all thumbnails of all displayed products are generated **before** the page is displayed.

In my opinion this is a big architectural mistake.
The user has to wait until tens of thumbnails are generated, before the first byte arrives, even though in the end for each product only one size is displayed.

#### Describe the solution you'd like

The HTML should be sent directly to the client regardless of the existence of the thumbnail images. The images should be resized on demand. No use to generate 5 thumbnails for each image (5 is PS default, could be more for some templates), if we display even one.

Class `ImageRetriever` does not resize images. If an inexistent product thumbnail is requested, `/img/p/.htaccess` redirects to a file `resize.php` that contains all the logic to generate the missing thumbnail, using `ImageManager::resize()`.

#### Alternatives you've considered

Alternatives Prestashop offers are insufficient: Image resizing via *Image Settings*, *Regenerate Thumbnails* is just impossible. An automated task takes days. 

#### PR in development

**Comment by hibatallahAouadni** [comment]
Hello @lmeyer1 

Thank you for your suggestion. The Product Team will take it into consideration for future developments. 
Please be aware that there is no guarantee that this feature will be developed anytime soon. If this is important to you, we strongly suggest you consider [hiring a specialist](https://www.prestashop.com/en/experts) to help you create it. 

If you are willing to implement this feature into PrestaShop as a [Pull Request](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/), the teams will be happy to help you by ensuring that it is appropriate and coherent with the rest of the project. If that is the case, please let us know! 

Thank you

**Comment by lmeyer1** [comment]
I understand that this feature is not in the list of priorities of Prestashop. Nevertheless I think it is very important.
Since we are running it for a long time now under PS 1.6, I thought others may take advantage of it, if I properly migrate it to PS 1.7 and publish a pull request here.
